ndk/src/main/jni/bugsnag_ndk.c

```diff
@@ -0,0 +1,432 @@
+    crumb->type = BSG_CRUMB_MANUAL;
+  }
+
+  bsg_populate_crumb_metadata(env, crumb, metadata);
+  bsg_request_env_write_lock();
+  bugsnag_report_add_breadcrumb(&bsg_global_env->next_report, crumb);
+  bsg_release_env_write_lock();
+
+  free(crumb);
+  (*env)->ReleaseStringUTFChars(env, name_, name);
+  (*env)->ReleaseStringUTFChars(env, crumb_type, type);
+  (*env)->ReleaseStringUTFChars(env, timestamp_, timestamp);
+}
+
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_ndk_NativeBridge_updateAppVersion(JNIEnv *env,
+                                                           jobject _this,
+                                                           jstring new_value) {
+  if (bsg_global_env == NULL)
+    return;
+  char *value = new_value == NULL
+                    ? NULL
+                    : (char *)(*env)->GetStringUTFChars(env, new_value, 0);
+  bsg_request_env_write_lock();
+  bugsnag_report_set_app_version(&bsg_global_env->next_report, value);
+  bsg_release_env_write_lock();
+  (*env)->ReleaseStringUTFChars(env, new_value, value);
+}
+
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_ndk_NativeBridge_updateBuildUUID(JNIEnv *env,
+                                                          jobject _this,
+                                                          jstring new_value) {
+  if (bsg_global_env == NULL)
+    return;
+  char *value = new_value == NULL
+                    ? NULL
+                    : (char *)(*env)->GetStringUTFChars(env, new_value, 0);
+  bsg_request_env_write_lock();
+  bugsnag_report_set_build_uuid(&bsg_global_env->next_report, value);
+  bsg_release_env_write_lock();
+  (*env)->ReleaseStringUTFChars(env, new_value, value);
+}
+
+JNIEXPORT void JNICALL Java_com_bugsnag_android_ndk_NativeBridge_updateContext(
+    JNIEnv *env, jobject _this, jstring new_value) {
+  if (bsg_global_env == NULL)
+    return;
+  char *value = new_value == NULL
+                    ? NULL
+                    : (char *)(*env)->GetStringUTFChars(env, new_value, 0);
+  bsg_request_env_write_lock();
+  bugsnag_report_set_context(&bsg_global_env->next_report, value);
+  bsg_release_env_write_lock();
+  if (new_value != NULL) {
+    (*env)->ReleaseStringUTFChars(env, new_value, value);
+  }
+}
+
+JNIEXPORT void JNICALL
+Java_com_bugsnag_android_ndk_NativeBridge_updateInForeground(
+    JNIEnv *env, jobject _this, jboolean new_value, jstring activity_) {
+  if (bsg_global_env == NULL)
+    return;
+  char *activity = activity_ == NULL
+                       ? NULL
+                       : (char *)(*env)->GetStringUTFChars(env, activity_, 0);
+  bsg_request_env_write_lock();
+  bsg_global_env->next_report.app.in_foreground = (bool)new_value;
+  bsg_strncpy_safe(bsg_global_env->next_report.app.active_screen, activity,
+                   sizeof(bsg_global_env->next_report.app.active_screen));
+  if ((bool)new_value) {
+    time(&bsg_global_env->foreground_start_time);
+  } else {
+    bsg_global_env->foreground_start_time = 0;
```

Comment:
Added extra proofing, though I haven't come up with an adequate test case for how it would be triggered. Can you think of one?

Reply:
I can't think of a case where this would be triggered off the top of my head - just saw that it was a possibility if this path was chosen for whatever reason.